### PR TITLE
Enforce validation of the ou id of the users

### PR DESCRIPTION
**Pull request by darshanasbg**
### Purpose
$subject. With this PR, during the user creation or user update, its performing strict validation of the OU id.
* OU ID should be valid UUID
* OU ID should be a ID of valid/existing OU
* OU ID of the create request or update request should be match with the user schema. ie. OU specified in the user create/update request should be same or child of the OU specified in the user schema.

### Approach
This pull request introduces validation to ensure that a user's organization unit is both valid and matches the user type's configuration, improving data integrity in user management operations. The changes include new error types, additional validation logic in user creation and update flows, and enhancements to organization unit hierarchy checks. Comprehensive unit tests have also been added to verify these behaviors.

**Organization Unit Validation Enhancements:**

* Added a new method `IsParent` to the `OrganizationUnitServiceInterface` and its implementation, allowing checks if an organization unit is an ancestor of another. [[1]](diffhunk://#diff-3182e56acf2b9cccc52c6c066f60106472bc9d1ba69738c1c639621615bcf510R44) [[2]](diffhunk://#diff-3182e56acf2b9cccc52c6c066f60106472bc9d1ba69738c1c639621615bcf510R235-R267)
* Extended the user service with a `validateOrganizationUnitForUserType` method to ensure the provided organization unit ID is valid, exists, and matches the user type’s schema or is a descendant. This validation is now invoked during user creation and update. [[1]](diffhunk://#diff-445a73f157e02dcd0f3f72d27f80664cbcc4de25fd0a6db25089a092f3c87f6aR170-R173) [[2]](diffhunk://#diff-445a73f157e02dcd0f3f72d27f80664cbcc4de25fd0a6db25089a092f3c87f6aR351-R354) [[3]](diffhunk://#diff-445a73f157e02dcd0f3f72d27f80664cbcc4de25fd0a6db25089a092f3c87f6aR560-R634)

**Error Handling Improvements:**

* Introduced new error constants: `ErrorInvalidOrganizationUnitID` and `ErrorOrganizationUnitMismatch` for more specific feedback when organization unit validation fails.

**Testing and Mocking:**

* Added comprehensive unit tests for the new organization unit hierarchy logic and validation behaviors, including edge cases and error handling.
* Updated organization unit and user service mocks to support the new `IsParent` method and validation logic.

**Test Consistency:**

* Refactored user service tests to consistently use a constant for user type, improving maintainability and readability. [[1]](diffhunk://#diff-0c0db27826059e36e41de3823032e63d8f80f10ec692af01196976ea5ccf5745R28-R42) [[2]](diffhunk://#diff-0c0db27826059e36e41de3823032e63d8f80f10ec692af01196976ea5ccf5745L52-R56) [[3]](diffhunk://#diff-0c0db27826059e36e41de3823032e63d8f80f10ec692af01196976ea5ccf5745L64-R68) [[4]](diffhunk://#diff-0c0db27826059e36e41de3823032e63d8f80f10ec692af01196976ea5ccf5745L74-R78) [[5]](diffhunk://#diff-0c0db27826059e36e41de3823032e63d8f80f10ec692af01196976ea5ccf5745L86-R90) [[6]](diffhunk://#diff-0c0db27826059e36e41de3823032e63d8f80f10ec692af01196976ea5ccf5745L100-R104) [[7]](diffhunk://#diff-0c0db27826059e36e41de3823032e63d8f80f10ec692af01196976ea5ccf5745L112-R116) [[8]](diffhunk://#diff-0c0db27826059e36e41de3823032e63d8f80f10ec692af01196976ea5ccf5745L122-R126) [[9]](diffhunk://#diff-0c0db27826059e36e41de3823032e63d8f80f10ec692af01196976ea5ccf5745L134-R143) [[10]](diffhunk://#diff-0c0db27826059e36e41de3823032e63d8f80f10ec692af01196976ea5ccf5745L153-R157) [[11]](diffhunk://#diff-0c0db27826059e36e41de3823032e63d8f80f10ec692af01196976ea5ccf5745L163-R172) [[12]](diffhunk://#diff-0c0db27826059e36e41de3823032e63d8f80f10ec692af01196976ea5ccf5745L178-R182)

### Related Issues
- N/A

### Related PRs
- N/A

### Checklist
- [ ] Followed the contribution guidelines.
- [ ] Manual test round performed and verified.
- [ ] Documentation provided. (Add links if there are any)
- [ ] Tests provided. (Add links if there are any)
    - [ ] Unit Tests
    - [ ] Integration Tests

### CODE REVIEW

File: backend/internal/user/service.go [hunk 1]
```diff
@@ -549,6 +557,81 @@ func (us *userService) ValidateUserIDs(userIDs []string) ([]string, *serviceerro
...
+	}
+
+	if us.ouService == nil {
+		logger.Error("Organization unit service is not configured for user operations")
+		return &ErrorInternalServerError
+	}
+
+	exists, svcErr := us.ouService.IsOrganizationUnitExists(organizationUnitID)
+	if svcErr != nil {
+		logger.Error("Failed to verify organization unit existence",
```
Comment (ThaminduDilshan) [inline backend/internal/user/service.go]: Shall we handle client errors here? There can be an invalid ouId in the user creation request

File: backend/internal/user/service_test.go [hunk 2]
```diff
@@ -25,17 +25,21 @@ import (
 	"github.com/stretchr/testify/mock"
 	"github.com/stretchr/testify/require"
 
+	// oupkg "github.com/asgardeo/thunder/internal/ou"
```
[the remark this answers is not in the document]
Comment (darshanasbg) [reply backend/internal/user/service_test.go]: Fixed with https://github.com/asgardeo/thunder/compare/76dfd9e64231736b46c89bb8e4a9629f20dc56c9..8b592b3e4f6e29f416c0bf70c099730ae1235f60

File: tests/integration/flowauthn/githubauth_test.go [hunk 3]
```diff
@@ -280,9 +281,9 @@ func (ts *GithubAuthFlowTestSuite) TestGithubAuthFlowCompleteSuccess() {
 		completeFlowStep.Assertion,
 		githubAuthTestAppID,
 		githubUserSchema.Name,
-		testOUID,
-		testOUName,
-		testOUHandle,
+		githubAuthTestOU.ID,
+		githubUserSchema.Name,
```
[the remark this answers is not in the document]
Comment (darshanasbg) [reply tests/integration/flowauthn/githubauth_test.go]: Fixed with https://github.com/asgardeo/thunder/compare/76dfd9e64231736b46c89bb8e4a9629f20dc56c9..8b592b3e4f6e29f416c0bf70c099730ae1235f60

File: backend/internal/user/service.go (hunk 1 above)
Comment (darshanasbg) [reply backend/internal/user/service.go]: Fixed with https://github.com/asgardeo/thunder/compare/8b592b3e4f6e29f416c0bf70c099730ae1235f60..85d5594fe7c8859941b8719dad8e6b0af13ab3b0

File: backend/internal/ou/service.go [hunk 4]
```diff
@@ -231,6 +232,39 @@ func (ous *organizationUnitService) IsOrganizationUnitExists(id string) (bool, *
...
+			}
+			logger.Error("Failed to traverse organization unit hierarchy", log.Error(err))
+			return false, &ErrorInternalServerError
+		}
+
+		currentParent = parentOU.Parent
+	}
+
+	return false, nil
+}
```
[the remark this answers is not in the document]
Comment (darshanasbg) [reply backend/internal/ou/service.go:266]: This is already ensured in the createOU, updateOU methods.. So no need to validate it again here.

**Review by senthalan — DISMISSED** [review]

**Review by ThaminduDilshan — DISMISSED** [review]
